routing/index_graph.hpp

```diff
@@ -29,23 +32,130 @@ class JointEdge final
+  }
+
+  bool operator==(DirectedEdge const & rhs) const
+  {
+    return m_from == rhs.m_from && m_to == rhs.m_to && m_featureId == rhs.m_featureId;
+  }
+
+  Joint::Id m_from = Joint::kInvalidId;
+  Joint::Id m_to = Joint::kInvalidId;
+  uint32_t m_featureId = 0;
+};
+
+inline bool IsCompatable(DirectedEdge const & ingoing, DirectedEdge const & outgoing)
+{
+  return ingoing.m_to == outgoing.m_from;
+}
+
+class RestrictionInfo final
+{
+public:
+  RestrictionInfo() = default;
+  RestrictionInfo(Joint::Id from, Joint::Id to, Joint::Id center,
+                  uint32_t fromFeatureId, uint32_t toFeatureId)
+    : m_from(from), m_to(to), m_center(center),
+      m_fromFeatureId(fromFeatureId), m_toFeatureId(toFeatureId)
+  {
+  }
+
+  RestrictionInfo(DirectedEdge const & ingoing, DirectedEdge const & outgoing)
+    : m_from(ingoing.m_from), m_to(outgoing.m_to), m_center(ingoing.m_to),
+      m_fromFeatureId(ingoing.m_featureId), m_toFeatureId(outgoing.m_featureId)
+  {
+    CHECK_EQUAL(ingoing.m_to, outgoing.m_from, ());
+  }
+
+  bool operator<(RestrictionInfo const & rhs) const
+  {
+    if (m_from != rhs.m_from)
+      return m_from < rhs.m_from;
+
+    if (m_to != rhs.m_to)
+      return m_to < rhs.m_to;
+
+    if (m_center != rhs.m_center)
+      return m_center < rhs.m_center;
+
+    if (m_fromFeatureId != rhs.m_fromFeatureId)
+      return m_fromFeatureId < rhs.m_fromFeatureId;
+
+    return m_toFeatureId < rhs.m_toFeatureId;
+  }
+
+  bool operator==(RestrictionInfo const & rhs) const { return !(*this < rhs || rhs < *this); }
+
+  bool operator!=(RestrictionInfo const & rhs) const { return !(*this == rhs); }
+
+  pair<DirectedEdge, DirectedEdge> ToEdges()
+  {
+    return make_pair(DirectedEdge(m_from, m_center, m_fromFeatureId),
+                     DirectedEdge(m_center, m_to, m_toFeatureId));
+  }
+
+  Joint::Id m_from = Joint::kInvalidId;
+  Joint::Id m_to = Joint::kInvalidId;
+  Joint::Id m_center = Joint::kInvalidId;
+  uint32_t m_fromFeatureId = 0;
+  uint32_t m_toFeatureId = 0;
 };
 
 class IndexGraph final
 {
 public:
+  static uint32_t const kStartFakeFeatureIds = 1024 * 1024 * 1024;
+
   IndexGraph() = default;
   explicit IndexGraph(unique_ptr<GeometryLoader> loader, shared_ptr<EdgeEstimator> estimator);
 
   // Creates edge for points in same feature.
   void GetDirectedEdge(uint32_t featureId, uint32_t pointFrom, uint32_t pointTo, Joint::Id target,
                        bool forward, vector<JointEdge> & edges);
-  void GetNeighboringEdges(RoadPoint const & rp, bool isOutgoing, vector<JointEdge> & edges);
+  void GetNeighboringEdges(RoadPoint const & rp, bool isOutgoing, bool graphWithoutRestrictions,
+                           vector<JointEdge> & edges);
 
   // Put outgoing (or ingoing) egdes for jointId to the 'edges' vector.
-  void GetEdgeList(Joint::Id jointId, bool isOutgoing, vector<JointEdge> & edges);
+  void GetEdgeList(Joint::Id jointId, bool isOutgoing, bool graphWithoutRestrictions,
```

Comment:
Pls have a look at IndexGraph::ApplyRestrictionPrepareData() method. We need to look form ingoing and outgoing edges on original graph in it.